### ISSUE TITLE
cgo: static linking a C library doesn't work.

### ISSUE THREAD

**Issue by chakrit** [constabulary/gb#331]
This is the library I'm building: https://github.com/hyperworks/go-barcode To build the `libzxing.a` file run the `build-zxing.sh` script in the repo folder.

To test, you can try vendoring that one into a GB project and just do `import _ "github.com/hyperworks/go-barcode"` and `gb build` the main project.

The build fails with a lot of errors that looks like `cgo` is trying to read the `libzxing.a` file as a text file (instead of as a library to link against.)

**Comment by chakrit** [comment]
Weird, I tried this on a fresh project and now it complains about a malformed `#cgo` directive instead.

```
FATAL command "build" failed: failed to resolve import path "m": /private/tmp/x/vendor/src/github.com/hyperworks/go-barcode/bridge.go: malformed #cgo argument: -I${SRCDIR}/zxing-cpp/core/src
```

**Comment by davecheney** [comment]
@chakrit thanks for your error report, the first error is most likely a dup of #307

**Comment by leifmadsen** [comment]
I have linked to an issue I originally filed with hlandau/service tag 1.0.7. It is very easy to reproduce with gb, and works fine with `go build`. I'm not sure if this is the same issue, but it is related to compiling C code embedded in a Go project.

**Comment by davecheney** [comment]
@leifmadsen can you try 0.1.1 or HEAD, you should get a different error (I'm still working on a fix)

**Comment by leifmadsen** [comment]
@davecheney sure I'll give it a shot in a few and report back

**Comment by leifmadsen** [comment]
Here is the output. Oddly enough, when I ran a `gb build` after updating gb, it seemed to compile fine. On the second build though, it provided the following output, which doesn't look any different unfortunately.

I ran an update for `gb` and verified I was running HEAD (master).

```sh
gb build
gopkg.in/hlandau/service.v1/exepath
gopkg.in/hlandau/service.v1/daemon/bansuid
gopkg.in/hlandau/service.v1/daemon/pidfile
gopkg.in/hlandau/service.v1/sdnotify
gopkg.in/hlandau/service.v1/daemon/caps
gopkg.in/hlandau/service.v1/daemon/setuid
/tmp/gb781477557/gopkg.in/hlandau/service.v1/daemon/_cgo_gotypes.go:3: package setuid; expected caps
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/caps/caps_linux.go:70[/tmp/gb781477557/gopkg.in/hlandau/service.v1/daemon/caps_linux.cgo1.go:17]: undefined: _Cfunc_hasanycap
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/caps/caps_linux.go:74[/tmp/gb781477557/gopkg.in/hlandau/service.v1/daemon/caps_linux.cgo1.go:23]: undefined: _Cfunc_dropcaps
gopkg.in/hlandau/service.v1/passwd
FATAL command "build" failed: exit status 1
```

**Comment by leifmadsen** [comment]
OK, actually now I do get a different message. I had to clear out the `/tmp/gbXXXXXXXX` dirs and then built again. It built fine after I did that. I then ran a `gb vendor update --all`.

I got not output, so I'm not sure if my `git diff` is just showing me the changes I had from previous to that command. I can confirm that the `service` vendor did not update though (same revision in the manifest).

Upon a new `gb build`, I got the following error:

```sh
$ gb build
github.com/go-yaml/yaml
github.com/VividCortex/siesta
github.com/pborman/uuid
gopkg.in/hlandau/service.v1/exepath
gopkg.in/hlandau/service.v1/daemon/bansuid
gopkg.in/hlandau/service.v1/daemon/pidfile
gopkg.in/hlandau/service.v1/sdnotify
/tmp/gb156964802/gopkg.in/hlandau/service.v1/daemon/_cgo_main.o: file not recognized: File truncated
collect2: error: ld returned 1 exit status
gopkg.in/hlandau/service.v1/passwd
gopkg.in/hlandau/service.v1/daemon/caps
github.com/ErikDubbelboer/gspt
config
FATAL command "build" failed: # gcc -fPIC -fmessage-length=0 -pthread -o /tmp/gb156964802/gopkg.in/hlandau/service.v1/daemon/_cgo_.o /tmp/gb156964802/gopkg.in/hlandau/service.v1/daemon/_cgo_main.o /tmp/gb156964802/gopkg.in/hlandau/service.v1/daemon/_cgo_export.o /tmp/gb156964802/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo2.o: exit status 1
```

The funny thing, is that now I've gone back to trying to reproduce the issue, and it's funny...

If I do this:

```
$ rm -rf /tmp/gb*
$ gb build
```

It all completes fine and seems to build. If I then re-run `gb build` it just returns as nothing appears to have changed.

Now if I do a `gb vendor update --all` (even though I just did this, and thus nothing should technically have changed), and then I run `gb build`, it fails again. This time with the following output:

```sh
 gb build
github.com/go-yaml/yaml
github.com/VividCortex/siesta
github.com/pborman/uuid
gopkg.in/hlandau/service.v1/exepath
gopkg.in/hlandau/service.v1/daemon/bansuid
gopkg.in/hlandau/service.v1/daemon/pidfile
gopkg.in/hlandau/service.v1/sdnotify
gopkg.in/hlandau/service.v1/daemon/caps
gopkg.in/hlandau/service.v1/daemon/setuid
gopkg.in/hlandau/service.v1/passwd
/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/_cgo_gotypes.go:3: package caps; expected setuid
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:43[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:13]: undefined: _Cfunc_csetuid
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:43[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:13]: undefined: _Ctype_uid_t
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:51[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:23]: undefined: _Cfunc_csetgid
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:51[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:23]: undefined: _Ctype_gid_t
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:63[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:39]: undefined: _Cfunc_csetresgid
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:63[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:39]: undefined: _Ctype_gid_t
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:71[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:49]: undefined: _Cfunc_csetresuid
/home/lmadsen/src/avoxi/api-proxy/vendor/src/gopkg.in/hlandau/service.v1/daemon/setuid/setuid_linux.go:71[/tmp/gb162270783/gopkg.in/hlandau/service.v1/daemon/setuid_linux.cgo1.go:49]: too many errors
github.com/ErikDubbelboer/gspt
config
FATAL command "build" failed: exit status 1
```

So it appears to be inconsistent in how it is using the cache perhaps? Let me know if there are any other tests I can run or output I can provide.

**Comment by davecheney** [comment]
Thanks for your response.

To clarify the change that landed in 0.1.1 didn't fix the issue, only the
panic during builds.

**Comment by leifmadsen** [comment]
No worries just wanted to provide as much info as possible. Wasn't sure if the issue was easily reproducible or if there was any new info there.

Let me know if there is anything else I can test and provide.

**Comment by davecheney** [comment]
@leifmadsen in your original report you mentioned the go-barcode library, which used cgo and c++. The issues with the services.v1/.../setuid library seem to be unrelated.

Are they related

**Comment by davecheney** [comment]
@leifmadsen thanks, i've replicated the problems with truncated .o files. I believe I know the problem. As this is a slightly different issue i'll raise issue to track it.

**Comment by davecheney** [comment]
@leifmadsen i'm pretty sure I fixed your issue in #372, @chakrit i'm still working on yours, although I have added some support for compiling .cpp files in #368

**Comment by leifmadsen** [comment]
@davecheney sorry, they are likely not related. I started here because it looked pretty close at first glance, then probably overloaded this issue. Should likely have just started with a new one because issue numbers are cheap :)

Thanks!
